### PR TITLE
WiFi Connector 2.0

### PR DESCRIPTION
**Pull request by jmvdwerf**
Most important changes in this feature:

- component Event Handlers uses the new event handling system. A handler can either be registered for a specific event_base (e.g., ```WIFI_EVENT``` or ```IP_EVENT```), or none (```ESP_EVENT_ANY_BASE```). The main loop should register the default event_loop first (i.e., call     ```ESP_ERROR_CHECK( esp_event_loop_create_default() );``` in the ```app_main``` method.

- WifiConnector uses the new IDF event passing system. The component connects to a SSID provided to the station. If no SSID is set, or it cannot connect to the specified AP, it becomes an AP itself.

### CODE REVIEW

File: main/main.cpp [hunk 1]
```diff
@@ -11,34 +11,37 @@
 
 #include "event_handlers.hpp"
 
-#include "wifiswitcher.hpp"
+#include "wificonnector.hpp"
+#include "wifisettings.h"
 
 extern "C" {
 
+    /*
```
Comment (fuegas) [inline main/main.cpp]: Please remove dead code.

File: components/wificonnector/include/wifiap.hpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+class WifiAP : public WifiBase
+{
+public:
+  WifiAP();
+
+  void ssid(std::string);
+  void password(std::string);
+
+  void authmode(wifi_auth_mode_t);
```
Comment (fuegas) [inline components/wificonnector/include/wifiap.hpp]: Why are you exposing these settings (auth_mode, channel and max_connections)?

File: components/wificonnector/include/wifiap.hpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#ifndef WIFIAP_HPP_
+#define WIFIAP_HPP_
+
+#include "wifibase.hpp"
```
Comment (fuegas) [inline components/wificonnector/include/wifiap.hpp]: If the name of a class is "WifiBase", please name the file "wifi_base.cpp".

File: components/event_handlers/event_handlers.cpp [hunk 4]
```diff
@@ -1,34 +1,83 @@
...
+        esp_event_handler_unregister(
+          ESP_EVENT_ANY_BASE,
+          ESP_EVENT_ANY_ID,
+          handle_event);
       }
+      return;
+    }
+    if (handlers.find(event_base) == handlers.end())
+    {
+      // handler was never registered for this event base!
```
Comment (fuegas) [inline components/event_handlers/event_handlers.cpp]: Nitpicking: Please start sentences with a Capital.

File: components/event_handlers/event_handlers.cpp [hunk 5]
```diff
@@ -1,34 +1,83 @@
...
 #include "event_handlers.hpp"
 
 namespace EventHandlers
 {
-  // List of handlers
-  std::list<EventHandler*> handlers;
 
-  void add(EventHandler *handler)
+  std::unordered_map<esp_event_base_t, std::list<EventHandler*>> handlers;
+  std::list<EventHandler*> allHandlers;
```
Comment (fuegas) [inline components/event_handlers/event_handlers.cpp]: Do you really need to keep a separate list with all handler? For example, using the lists the way you do now, does that cause events to be called twice? You're calling events on `allHandlers` and on `handlers`.

**Review by fuegas — CHANGES_REQUESTED** [review]
Good refactor. The names `handlers` and `allHandlers` confuse me though...  Apparently `handlers` is a subset or something? As it isn't described in a comment it's hard to figure out what they are for.

File: components/event_handlers/event_handlers.cpp (hunk 5 above)
Comment (jmvdwerf) [reply components/event_handlers/event_handlers.cpp]: I renamed ```allHandlers``` to ```genericHandlers```, as it stores a different set of handlers. Also added a comment explaining the difference.

Short summary: handlers can either be registered to a specific ```event_base```, or to ```any```. The former is implemented as the map ```handlers```, the latter as the list ```genericHandlers```.

File: components/event_handlers/event_handlers.cpp (hunk 4 above)
Comment (jmvdwerf) [reply components/event_handlers/event_handlers.cpp]: Fixed!

File: components/wificonnector/include/wifiap.hpp (hunk 3 above)
Comment (jmvdwerf) [reply components/wificonnector/include/wifiap.hpp]: Done!

File: components/wificonnector/include/wifiap.hpp (hunk 2 above)
Comment (jmvdwerf) [reply components/wificonnector/include/wifiap.hpp]: The main reason is that I don't know yet who will "own" the settings, and where and how these are stored and retrieved. If that will be the task of this component as well, it is not required to expose any of them anymore.

File: main/main.cpp (hunk 1 above)
Comment (jmvdwerf) [reply main/main.cpp]: Done!

**Review by jmvdwerf — COMMENTED** [review]
Done!

File: components/event_handlers/event_handlers.cpp [hunk 6]
```diff
@@ -1,34 +1,68 @@
...
-        return err;
-      }
+      genericHandlers.remove(handler);
+      return;
     }
 
-    return ESP_OK;
+    handlers[event_base].remove(handler);
   }
+
```
Comment (fuegas) [inline components/event_handlers/event_handlers.cpp]: Nitpick: Leftover whitespace.

File: components/event_handlers/include/event_handlers.hpp [hunk 7]
```diff
@@ -11,10 +11,9 @@
  */
 namespace EventHandlers
 {
-  void add(EventHandler*);
-  void remove(EventHandler*);
-
-  esp_err_t handle_event(void*, system_event_t*);
+  void init();
```
Comment (fuegas) [inline components/event_handlers/include/event_handlers.hpp]: Nitpick: Please use `void init(void);`

File: components/wifi_connector/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,13 @@
+set(COMPONENT_ADD_INCLUDEDIRS include)
+set(COMPONENT_SRCS
+        "wifi_base.cpp"
```
Comment (fuegas) [inline components/wifi_connector/CMakeLists.txt]: Nitpick: Please use 2 spaces for indentation.

File: components/wifi_connector/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+        "wifi_ap.cpp"
+        "wifi_station.cpp"
+        "wifi_connector.cpp"
+)
+
+set(COMPONENT_REQUIRES
+        event_handlers
```
Comment (fuegas) [inline components/wifi_connector/CMakeLists.txt]: Nitpick: ditto.

File: components/wifi_connector/include/wifi_ap.hpp [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+  void connect();
+
+  void handle_event(esp_event_base_t, int32_t, void *);
+private:
+  wifi_config_t cfg_ap_ = {};
+
+};
+
+
```
Comment (fuegas) [inline components/wifi_connector/include/wifi_ap.hpp]: Nitpick: Double newline.

File: components/wifi_connector/include/wifi_connector.hpp [hunk 11]
```diff
@@ -0,0 +1,40 @@
+#ifndef WIFI_CONNECTOR_HPP_
+#define WIFI_CONNECTOR_HPP_
+
+#include "event_handler.hpp"
+
```
Comment (fuegas) [inline components/wifi_connector/include/wifi_connector.hpp:5]: Nitpick: These includes are our own code, so this newline isn't needed.

File: components/wifi_connector/wifi_ap.cpp [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+#include "event_handlers.hpp"
+
+WifiAP::WifiAP()
+{
+  cfg_ap_ = {};
+  cfg_ap_.ap = {};
+  cfg_ap_.ap.ssid_hidden = 0;
+  cfg_ap_.ap.authmode = WIFI_AUTH_WPA_WPA2_PSK;
+
+  this->max_connections(4);
```
Comment (fuegas) [inline components/wifi_connector/wifi_ap.cpp:11]: Is there a reason to limit the number of allowed connections to a AP?

If there is no need to be able to change it later, you can set it directly in the config above as well.
```c++
cfg_ap_.ap.max_connection = 4;
```

Then the whole maxconnections method can be removed I guess?

File: components/wifi_connector/wifi_ap.cpp [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+void WifiAP::ssid(std::string ssid)
+{
+  if (ssid.length() < 32) {
+    std::copy(ssid.begin(), ssid.end(), cfg_ap_.ap.ssid);
+    ssid_ = ssid;
+  }
+  else
+  {
+    ssid_ = "";
```
Comment (fuegas) [inline components/wifi_connector/wifi_ap.cpp]: Is an empty string a correct fallback if the given string is longer than 32 chars? Would a neater solution be to create an AP with the allowed length of the provided SSID?

File: components/wifi_connector/wifi_ap.cpp [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+  }
+}
+
+void WifiAP::password(std::string password)
+{
+  if (password.length() < 64) {
+    std::copy(password.begin(), password.end(), cfg_ap_.ap.password);
+  }
```
Comment (fuegas) [inline components/wifi_connector/wifi_ap.cpp:34]: If the SSID is too long, you set it to `""`, for the password there is no fallback behaviour, is this intended? What happens if the password is too long? Do we have a public (non-password-protected) AP then?

File: components/wifi_connector/wifi_connector.cpp [hunk 15]
```diff
@@ -0,0 +1,80 @@
+#include "wifi_connector.hpp"
+#include "event_handlers.hpp"
```
Comment (fuegas) [inline components/wifi_connector/wifi_connector.cpp:1]: Nitpick: Please order lexicographically.

File: components/wifi_connector/wifi_station.cpp [hunk 16]
```diff
@@ -0,0 +1,88 @@
+#include "wifi_station.hpp"
+
```
Comment (fuegas) [inline components/wifi_connector/wifi_station.cpp]: Nitpick: This newline can be removed, and the includes ordered lexicographically.

File: components/wifi_connector/wifi_station.cpp [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+void WifiStation::ssid(std::string ssid)
+{
+  if (ssid.length() < 32) {
+    std::copy(ssid.begin(), ssid.end(), cfg_station_.sta.ssid);
+    ssid_ = ssid;
+  }
+  else
+  {
+    ssid_ = "";
```
Comment (fuegas) [inline components/wifi_connector/wifi_station.cpp]: Same question here as at the SSID of the AP code.

File: components/wifi_connector/wifi_station.cpp [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+  }
+}
+
+void WifiStation::password(std::string password)
+{
+  if (password.length() < 64) {
+    std::copy(password.begin(), password.end(), cfg_station_.sta.password);
+  }
```
Comment (fuegas) [inline components/wifi_connector/wifi_station.cpp:39]: Same question here as at the password of the AP code.

**Review by fuegas — APPROVED** [review]
I'll approve the code as I do not see big flaws in it. But please process the comments into a new PR.

**Review by jmvdwerf — COMMENTED** [review]
I made the changes as you requested. Password and SSID are now capped at 32 and 64, resp. if they are too long.

**Pull request merged**